File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonAsyncTaskShouldNotReturnNull.cs
using System;
using System.Threading.Tasks;

namespace Tests.Diagnostics
{
    public class CompliantUseCases
    {
        public async Task GetTaskAsync1()
        {
        }

        public async Task<object> GetTaskAsync2()
        {
            return null; // async
        }

        public Task GetTask1()
        {
            return Task.FromResult(true);
        }

        public Task GetTask2()
        {
            var x = new Task<object>(() => null); // Compliant

            return Task.Delay(0);
        }

        public Task Foo { get; set; }

        public Task<object> GetFooAsync()
        {
            return Task.Run(() =>
            {
                if (false)
                {
                    return new object();
                }
                else
                {
                    return null;
                }
            });
        }

        public Task GetBar()
        {
            Func<Task> func = () => null;

            return func(); // False negative
        }
    }

    public class NonCompliantUseCases
    {
        public Task GetTask1()
        {
            return null; // Noncompliant {{Do not return null from this method, instead return 'Task.FromResult<T>(null)', 'Task.CompletedTask' or 'Task.Delay(0)'.}}
//                 ^^^^
        }

        public Task<object> GetTask2()
        {
            return null; // Noncompliant
        }

        public Task GetTaskAsync3(int a)
        {
            if (a > 42)
            {
                return null; // Noncompliant
            }

            return Task.Delay(0);
        }

        public Task<string> GetTask4() => null; // Noncompliant

        public Task<string> GetTask5(int a)
        {
            if (a > 0)
            {
                return null; // Noncompliant
            }
            else
            {
                return null; // Noncompliant
            }
        }

        public Task GetTask6()
        {
            return (null); // Noncompliant
        }

        public Task GetTask7(bool condition)
        {
            return condition ? Task.FromResult(5) : null; // Should be non-compliant
        }

        private Task foo;
        public Task Foo
        {
            get
            {
                return null; // Noncompliant
            }
            set
            {
                foo = value;
            }
        }

        public Task<int> AgeAsync
        {
            get => null; // Noncompliant
        }

        // See https://github.com/SonarSource/sonar-dotnet/issues/1845
        public Task<object> GetObjectTask()
        {
            object GetObject()
            {
                return null;
            }

            return Task.FromResult(GetObject());
        }

        public Task<object> SomeMethodAsync()
        {
            return this.SomeOtherMethodAsync(
                async input =>
                {
                    await Task.Yield();

                    return null;
                });
        }

        public Task<object> SomeMethodNonAsync()
        {
            return this.SomeOtherMethodAsync(
                input =>
                {
                    return null; // Noncompliant
                });
        }

        public Task<object> SomeMethodParenthesizedAsync()
        {
            return this.SomeOtherMethodAsync(
                async (input) =>
                {
                    await Task.Yield();

                    return null;
                });
        }

        public Task<object> SomeMethodParenthesizedNonAsync()
        {
            return this.SomeOtherMethodAsync(
                (input) =>
                {
                    return null; // Noncompliant
                });
        }

        private Task<object> SomeOtherMethodAsync(Func<object, Task<object>> function) => function.Invoke(null);

        public Task<object> GetBar()
        {
            return Add(1, 2);

            static Task<object> Add(int left, int right)
            {
                return null; // Noncompliant
            }

            static Task<object> GetTaskAsync2() => null; // Noncompliant
        }
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonAsyncTaskShouldNotReturnNull.vb
Imports System
Imports System.Threading.Tasks

Namespace Tests.Diagnostics
    Public Class CompliantUseCases

        Public Async Function GetTaskAsync2() As Task(Of Object)
            Return Nothing
        End Function

        Public Function GetTask1() As Task
            Return Task.FromResult(True)
        End Function

        Public Function GetTask2() As Task
            Dim x = New Task(Of Object)(Function() Nothing)
            Return Task.Delay(0)
        End Function

        Public Property Foo As Task

        Public Function GetFooAsync() As Task(Of Object)
            Return Task.Run(Function()
                                If False Then
                                    Return New Object()
                                Else
                                    Return Nothing
                                End If
                            End Function)
        End Function

        Public Function GetBar() As Task
            Dim func As Func(Of Task) = Function() Nothing
            Return func()
        End Function
    End Class

    Public Class NonCompliantUseCases
        Public Function GetTask1() As Task
            Return Nothing ' Noncompliant {{Do not return null from this method, instead return 'Task.FromResult(Of T)(Nothing)', 'Task.CompletedTask' or 'Task.Delay(0)'.}}
'                  ^^^^^^^
        End Function

        Public Function GetTask2() As Task(Of Object)
            Return Nothing ' Noncompliant
        End Function

        Public Function GetTaskAsync3(ByVal a As Integer) As Task
            If a > 42 Then
                Return Nothing ' Noncompliant
            End If

            Return Task.Delay(0)
        End Function

        Public Function GetTask5(ByVal a As Integer) As Task(Of String)
            If a > 0 Then
                Return Nothing ' Noncompliant
            Else
                Return Nothing ' Noncompliant
            End If
        End Function

        Public Function GetTask6() As Task
            Return (Nothing) ' Noncompliant
        End Function

        Public Function GetTask7(ByVal condition As Boolean) As Task
            Return If(condition, Task.FromResult(5), Nothing) ' Noncompliant
        End Function

        Public Property Foo As Task
            Get
                Return Nothing ' Noncompliant
            End Get
            Set(ByVal value As Task)
                foo = value
            End Set
        End Property
    End Class
End Namespace
